support-workers/src/main/scala/com/gu/zuora/subscriptionBuilders/DigitalSubscriptionBuilder.scala

```diff
@@ -82,15 +83,20 @@ object DigitalSubscriptionBuilder {
       initialTerm = initialTerm
     )
 
-    val withRedemptionCode = generateRedemptionCode(readerType, purchase.billingPeriod, subscriptionData)
-    val withPromoApplied = applyPromoCode(purchase.promotionService, purchase.maybePromoCode, purchase.country, productRatePlanId, withRedemptionCode)
+    val withRedemptionCode = addRedemptionCodeIfGift(readerType, purchase.billingPeriod, subscriptionData)
+    val withPromoApplied = applyPromoCodeIfPresent(purchase.promotionService, purchase.maybePromoCode, purchase.country, productRatePlanId, withRedemptionCode)
 
     EitherT.fromEither[Future](withPromoApplied.left.map(Left.apply))
   }
 
-  def generateRedemptionCode(readerType: ReaderType, billingPeriod: BillingPeriod, subscriptionData: SubscriptionData) = {
+  def fromBillingPeriod(billingPeriod: BillingPeriod): GiftDuration = billingPeriod match {
+    case Annual => Gift12Month
+    case _ => Gift3Month
```

Comment:
in relality this is a partial function
annual->12month
quarter->3month
other->impossible presumably enforced by the validators in support-frontend backend

If this is changed or broken, we might silently create incorrect gifts which we may only realise when people who were expecting (6 month? recurring?) gifts find they are ending after only 3 months.  I wouldn't look forward to that repair operation.

Ideally we would only be able to represent valid gift durations in the state model ADT, meaning it wouldn't even compile if this assumption was broken or removed at the other end.

Failing that I think we should return an error e.g. Option at this point to at least fail before getting bad data into the zuora db.

Reply:
The thing is that all this really has nothing to do with the sub we are creating, it is just to generate the 'gd12' or 'gd03' part of the code. This was the reason why I was a bit wary of having information in the codes in the first place, because it is increasing the code complexity for something which is at best cosmetic and at worst misleading.

Reply:
oh yeah good point forgot about that - of course the original billing period object is the one that deals with zuora
In that case I'm less concerned as the user will still get what was paid for even if the gift-code isn't helpful to us.

Reply:
we could have a conversation about the code format - the code wasn't in my mind cosmetic, actually it was to make it easier for userhelp etc to understand what the code is without having to go into the system to look things up.  But I agree the default assumption seemed to be that the code has meaning in it, which I guess wasn't questioned enough in your opinion.

Reply:
I think it's marginal, I can see that there could be some value from a userhelp point of view but on the other hand it's another source of bugs so by default I would have gone for a YAGNI approach.